### DISCUSSION TITLE
What do I need to learn if I want to contribute to this open source project?

### DISCUSSION THREAD

**Discussion by Linhieng** [rolldown/rolldown#619] [General]
I've always wanted to implement my own packaging tool for personal use, and recently I came across the open-source project called "rolldown" which caught my interest. I believe I can follow this project and develop a simple packaging tool of my own. Eventually, I might even be able to submit a PR to contribute to this project.

Although I'm currently in the process of learning the basics of the Rust language, it doesn't deter me from envisioning the future.😀

However, I'm also aware that developing a project involves much more than just knowing the language. It requires additional knowledge such as data structures, algorithms, and even compiler theory. I'm curious if these areas of expertise will be needed in this project.

Finally, leaving this discussion here serves as a witness to my TODO list. Perhaps one day in the future, I will share my own packaging open-source project here.

---
<details>
  <summary>View original text (Chinese)</summary>
  
我一直想实现一个自己使用的打包工具，刚好最近看到了 rolldown 这个开源项目，我非常有兴趣。我觉得我可以跟随这个开源项目，自己实现一个简单的打包工具，然后慢慢地也许我也能提交为这个项目提一个 PR。

虽然目前我还在学习 rust 语言基础的过程中，但这并无法阻止我憧憬着未来。

但我也知道，开发一个项目，语言仅仅只是最基础的，在此之外还需要很多额外的知识，比如数据结构预算法、编译原理之类的。我想知道这个项目中是否会需要这些知识点？

最后，在这里留下一条 discussion，也算是一个 TODO 见证，也许未来的某天，我就会在这里分享我自己的打包开源项目了
</details>

**Comment by hyfdev** [comment]
Cool. Developing bundler isn't hard. You could start by reading code of esbuild and rollup. Especially for early version of rollup, it's a great example to show how a bundler works.

**Reply by Linhieng** [reply]
Thanks, I will make a note of that! :)

**Reply by j-aidle** [reply]
I'm also like to contribute to the rolldown project. As you I'm trying Rust and learning about Bundlers. we could help each other!

**Reply by Linhieng** [reply]
Mutual encouragement :) 

But right now my priority is to find a job and graduate successfully😀

**Comment by moushicheng** [comment]
Open source can contribute little by little, starting from their own will, there is no need to learn all and then engage, and even learn a little to engage a little

> 开源可以一点一点做贡献，从自己会的开始，没必要全学完再搞，甚至可以学一点搞一点

**Reply by moushicheng** [reply]
By the way, I'm hardreading the rolldown source code to understand the packager, and I'm not sure this is advisable\
> 顺带一提，我正在硬读rolldown源码来理解打包器，不知道这可不可取(?)

**Reply by Linhieng** [reply]
I agree with your point, but I also know that the efficiency of directly reading source code without understanding certain knowledge backgrounds is actually quite low. Not only does it result in low efficiency in time utilization, but it can also easily dampen your enthusiasm. That's why I'm leaving this discussion here, maybe one day I'll hook a frontend expert. 😁

<details>
<summary>Original Chinese</summary>
我同意你的观点，但我也知道，如果不了解某些知识背景，就直接阅读源代码的话，性价比其实是很低的。这不仅对时间的利用效率低，而且也容易打击你的激情。这也是为什么我会在这里留个 discussion，也许某天就钓到了个大佬呢。
</details>
